File: API/0848_Gold_Pullback_Strategy/CS/GoldPullbackStrategy.cs
using System;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Gold pullback strategy using EMA trend filter, MACD and TDI confirmation.
/// </summary>
public class GoldPullbackStrategy : Strategy
{
	private readonly StrategyParam<int> _emaFastLength;
	private readonly StrategyParam<int> _emaSlowLength;
	private readonly StrategyParam<int> _emaPullbackLength;
	private readonly StrategyParam<decimal> _slOffset;
	private readonly StrategyParam<DataType> _candleType;

	private ExponentialMovingAverage _emaFast;
	private ExponentialMovingAverage _emaSlow;
	private ExponentialMovingAverage _emaPullback;
	private MovingAverageConvergenceDivergence _macd;
	private RelativeStrengthIndex _rsi;
	private SimpleMovingAverage _tdiMa;
	private SimpleMovingAverage _tdiSignal;

	private decimal _longStop;
	private decimal _longTake;
	private decimal _shortStop;
	private decimal _shortTake;

	public GoldPullbackStrategy()
	{
		_emaFastLength = Param(nameof(EmaFastLength), 14)
			.SetGreaterThanZero()
			.SetDisplay("EMA Fast", "Fast EMA length", "Indicators")
			.SetCanOptimize(true);

		_emaSlowLength = Param(nameof(EmaSlowLength), 60)
			.SetGreaterThanZero()
			.SetDisplay("EMA Slow", "Slow EMA length", "Indicators")
			.SetCanOptimize(true);

		_emaPullbackLength = Param(nameof(EmaPullbackLength), 21)
			.SetGreaterThanZero()
			.SetDisplay("EMA Pullback", "Pullback EMA length", "Indicators")
			.SetCanOptimize(true);

		_slOffset = Param(nameof(SlOffset), 0.1m)
			.SetGreaterThan(0m)
			.SetDisplay("SL Offset", "Offset for stop calculation", "Risk")
			.SetCanOptimize(true);

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Candle type for strategy", "General");
	}

	public int EmaFastLength { get => _emaFastLength.Value; set => _emaFastLength.Value = value; }
	public int EmaSlowLength { get => _emaSlowLength.Value; set => _emaSlowLength.Value = value; }
	public int EmaPullbackLength { get => _emaPullbackLength.Value; set => _emaPullbackLength.Value = value; }
	public decimal SlOffset { get => _slOffset.Value; set => _slOffset.Value = value; }
	public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		=> [(Security, CandleType)];

	protected override void OnReseted()
	{
		base.OnReseted();
		_longStop = _longTake = 0m;
		_shortStop = _shortTake = 0m;
	}

	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		_emaFast = new ExponentialMovingAverage { Length = EmaFastLength };
		_emaSlow = new ExponentialMovingAverage { Length = EmaSlowLength };
		_emaPullback = new ExponentialMovingAverage { Length = EmaPullbackLength };
		_macd = new MovingAverageConvergenceDivergence
		{
			ShortPeriod = 5,
			LongPeriod = 34,
			SignalPeriod = 5
		};
		_rsi = new RelativeStrengthIndex { Length = 13 };
		_tdiMa = new SimpleMovingAverage { Length = 2 };
		_tdiSignal = new SimpleMovingAverage { Length = 7 };

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(_emaFast, _emaSlow, _emaPullback, _macd, _rsi, ProcessCandle).Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, _emaFast);
			DrawIndicator(area, _emaSlow);
			DrawIndicator(area, _emaPullback);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal emaFast, decimal emaSlow, decimal emaPullback, decimal macd, decimal macdSignal, decimal _, decimal rsiValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		var tdiMaValue = _tdiMa.Process(new DecimalIndicatorValue(_tdiMa, rsiValue)).ToDecimal();
		var tdiSignalValue = _tdiSignal.Process(new DecimalIndicatorValue(_tdiSignal, rsiValue)).ToDecimal();

		if (!_macd.IsFormed || !_tdiSignal.IsFormed || !_emaFast.IsFormed || !_emaSlow.IsFormed || !_emaPullback.IsFormed)
			return;

		var trendUp = emaFast > emaSlow;
		var trendDown = emaFast < emaSlow;
		var touchEma = candle.LowPrice <= emaPullback && candle.HighPrice >= emaPullback;
		var macdBuy = macd > macdSignal;
		var macdSell = macd < macdSignal;
		var tdiBuy = tdiMaValue > tdiSignalValue && rsiValue > 50m;
		var tdiSell = tdiMaValue < tdiSignalValue && rsiValue < 50m;

		var buySignal = trendUp && touchEma && macdBuy && tdiBuy;
		var sellSignal = trendDown && touchEma && macdSell && tdiSell;

		if (Position > 0)
		{
			if (candle.LowPrice <= _longStop || candle.HighPrice >= _longTake)
			{
				SellMarket(Position);
				_longStop = _longTake = 0m;
			}
		}
		else if (Position < 0)
		{
			if (candle.HighPrice >= _shortStop || candle.LowPrice <= _shortTake)
			{
				BuyMarket(Math.Abs(Position));
				_shortStop = _shortTake = 0m;
			}
		}

		if (buySignal && Position <= 0)
		{
			var volume = Volume + Math.Abs(Position);
			var sl = candle.LowPrice - SlOffset;
			var tp = candle.ClosePrice + (candle.ClosePrice - sl);
			BuyMarket(volume);
			_longStop = sl;
			_longTake = tp;
		}
		else if (sellSignal && Position >= 0)
		{
			var volume = Volume + Math.Abs(Position);
			var sl = candle.HighPrice + SlOffset;
			var tp = candle.ClosePrice - (sl - candle.ClosePrice);
			SellMarket(volume);
			_shortStop = sl;
			_shortTake = tp;
		}
	}
}

File: API/0848_Gold_Pullback/PY/gold_pullback_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage
from StockSharp.Algo.Strategies import Strategy

class gold_pullback_strategy(Strategy):
    """
    EMA crossover strategy.
    Buys when fast EMA crosses above slow EMA, sells when it crosses below.
    """

    def __init__(self):
        super(gold_pullback_strategy, self).__init__()
        self._fast_period = self.Param("FastPeriod", 120) \
            .SetDisplay("Fast Period", "Fast EMA period", "General")
        self._slow_period = self.Param("SlowPeriod", 450) \
            .SetDisplay("Slow Period", "Slow EMA period", "General")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(1))) \
            .SetDisplay("Candle Type", "Candle timeframe", "General")

        self._prev_fast = 0.0
        self._prev_slow = 0.0

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(gold_pullback_strategy, self).OnReseted()
        self._prev_fast = 0.0
        self._prev_slow = 0.0

    def OnStarted(self, time):
        super(gold_pullback_strategy, self).OnStarted(time)

        fast_ema = ExponentialMovingAverage()
        fast_ema.Length = self._fast_period.Value
        slow_ema = ExponentialMovingAverage()
        slow_ema.Length = self._slow_period.Value

        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(fast_ema, slow_ema, self._process_candle).Start()

    def _process_candle(self, candle, fast_val, slow_val):
        if candle.State != CandleStates.Finished:
            return

        if not self.IsFormedAndOnlineAndAllowTrading():
            return

        fast = float(fast_val)
        slow = float(slow_val)

        if self._prev_fast != 0.0 and self._prev_slow != 0.0:
            if self._prev_fast <= self._prev_slow and fast > slow:
                if self.Position <= 0:
                    self.BuyMarket()
            elif self._prev_fast >= self._prev_slow and fast < slow:
                if self.Position >= 0:
                    self.SellMarket()

        self._prev_fast = fast
        self._prev_slow = slow

    def CreateClone(self):
        return gold_pullback_strategy()
